Add controllers

svlgerasimov opened the pull request.

svlgerasimov committed:
feat: add user

svlgerasimov committed:
feat: доделать пользователей

svlgerasimov committed:
feat: добавить модели

svlgerasimov committed:
feat: добавить фичи Item, добавить ErrorHa

svlgerasimov committed:
fix: убрать валидацию id

svlgerasimov committed:
feat: добавить исключение для валидации

svlgerasimov committed:
feat: добавить логгирование
refactor: программная валидация в контроллерах

svlgerasimov committed:
codestyle: требования jls

svlgerasimov committed:
feat: добавить logbook и настройки уровня логгирования

svlgerasimov committed:
codestyle: убрать лишний закомментированный код

svlgerasimov committed:
codestyle: убрать реализованные TODO комментарии

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,44 @@
+
+public class ItemMapper {
+
+    private ItemMapper() {
```
(предложение)Можно добавить такую аннотацию
`@NoArgsConstructor(access = AccessLevel.PRIVATE)`
Чтобы создать приватный конструктор без параметров

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserMapper.java:
```diff
@@ -0,0 +1,43 @@
+
+public class UserMapper {
+
+    private UserMapper() {
```
(предложение)Можно добавить такую аннотацию
`@NoArgsConstructor(access = AccessLevel.PRIVATE)`
Чтобы создать приватный конструктор без параметров

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,25 @@
+import java.util.Objects;
+
+@Value
 public class ItemDto {
```
Можно было бы использовать аннотации валидации для того, чтобы избавиться от проверок в коде. Например, над полем с названием можно было бы использовать `@NotBlank`. Аналогично для описания)
Для available можно использовать `@NotNull`, так как она может иметь лишь два состояния и следовательно должно быть передано

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @PostMapping
+    public ItemDto add(@RequestBody ItemDto itemDto,
```
Теперь для того, чтобы предложенные аннотации в классе `ItemDto` стали учитывать, необходимо перед аргументом метода добавить `@Valid`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    private void patchValidateItemDto(ItemDto itemDto) {
```
Проверка не требуется, так как мы на уровне сервиса должны проверить, что имя не null и не пусто и не состоит из пробелов. И в случае, если оба условия - true, необходимо установить эти значения полученному из хранилища обновляемому объекту) Аналогично с описанием и available

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,73 @@
+        return itemService.patch(id, patchDto, userId);
+    }
+
+    private void fullValidateItemDto(ItemDto itemDto) {
```
Метод более не требуется, так как проверка происходит с помощью аннотаций валидации

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,44 @@
+
+    public static void patchFromDto(Item itemToPatch, ItemDto patchDto) {
+        String patchName = patchDto.getName();
+        if (Objects.nonNull(patchName)) {
```
не забываем проверить строку с помощью isBlank, так как она может и не быть null, но может быть пустой и состоящей из пробелов

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,44 @@
+            itemToPatch.setName(patchName);
+        }
+        String patchDescription = patchDto.getDescription();
+        if (Objects.nonNull(patchDescription)) {
```
не забываем проверить строку с помощью isBlank, так как она может и не быть null, но может быть пустой и состоящей из пробелов

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public List<ItemDto> search(String text) {
+        if (text.isEmpty()) {
```
Проверку хорошо было бы вынести на уровень контроллера, так как происходит проверка данных, сформированных не в результате работы бизнес логики

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public ItemDto add(ItemDto dto, long userId) {
+        User owner = userStorage.getById(userId)
```
Можно было бы эту логику вынести в отдельный приватный метод, так как она дублируется в двух методах)
А так же для соблюдения принципа единственной ответственности

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/controller/UserController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @PostMapping
+    public UserDto add(@RequestBody @Valid UserDto userDto) {
```
(предложение) Кстати, вместо @Valid можно использовать такую аннотацию @Validated(Create.class). Таким образом мы указываем, какую группу аннотаций следует учитывать. Create - это интерфейс, который ты создаешь самостоятельно, он будет пустым и служит лишь для указания группы) Аналогично следует поступить при обновлении, но там будет аннотация @Validated(Update.class)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/controller/UserController.java:
```diff
@@ -0,0 +1,56 @@
+        userService.remove(id);
+    }
+
+    private void fullValidateUserDto(UserDto userDto) {
```
Метод не требуется, так как валидация будет реализована с помощью аннотаций)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.Objects;
+
+@Value
+public class UserDto {
```
Так же и здесь, можно было бы использовать валидацию по группам)
Для имени:
`@NotNlank(groups = {Create.class})`
Для почты аннотацию Email следует учитываться для обновления и создания, так как, если она не будет передана вообще, то все будет окей, а если будет, тогда она будет проверена на соответствие почте, а вот аннотацию
NotBlank только для создания) Так как при обновлении передавать ее не обязательно)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Slf4j
+//@Validated
```
Если код не используется - следует его убрать

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/storage/ItemStorageInMemory.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public List<Item> getAll(User owner) {
```
(предложение)Для упрощения работы можно использовать еще одну мапу вида:
`private final Map<Long, List<Item>> userItemIndex = new LinkedHashMap<>();`
В ней ключами будут выступать идентификаторы пользователей, а в виде значений, будут их вещи) Таким образом, если нам потребуется получить все вещи пользователя, нам достаточно будет обратиться к этой мапе, без итерации по всем значениям другой)

При сохранении можно будет использовать такую запись)
`final List<Item> items = userItemIndex.computeIfAbsent(item.getOwner().getId(), k -> new ArrayList<>());`
Таким образом, благодаря методу computeIfAbsent будет создан новый объект ArrayList'a, если его нет в мапе по ключу, либо же вернется ссылка на него, если он будет найден) А далее мы уже сможем в него добавить новый предмет пользователя)

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил мелкие недочеты, которые следует исправить. Хорошего дня :) Если возникнут вопросы - пиши в слак, буду рад помочь!

svlgerasimov committed:
refactor: валидация через validation groups

svlgerasimov committed:
feat: добавить проверку NotBlank имени пользователя, убрать вывод email для пустого имени

svlgerasimov committed:
style: приватные конструкторы через lombok

svlgerasimov committed:
refactor: в ItemServiceImpl выделить получение владельца из репозитория в отдельный метод

svlgerasimov committed:
refactor: перенести проверку пустой строки поиска на уровень контроллера

svlgerasimov committed:
style: убрать ненужные аннотации и комментарии

svlgerasimov committed:
refactor: добавить в ItemStorageInMemory индекс для поиска по владельцу

svlgerasimov committed:
style: убрать лишние аннотации

svlgerasimov committed:
style: убрать лишние todo аннотации

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,46 @@
+
+    @GetMapping("/search")
+    public List<ItemDto> search(@RequestParam String text) {
+        return text.isEmpty() ? Collections.emptyList() : itemService.search(text);
```
Хорошо было бы проверить строку с помощью `isBlank`, так как `isEmpty` допускает строку, состоящую только из пробелов - за корректную)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,31 @@
+    Long id;
+
+    @NotBlank(groups = FullValidated.class)
+    @NullableNotBlank(groups = PatchValidated.class)
```
Классное решение! Мне нравится :)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/controller/UserController.java:
```diff
@@ -0,0 +1,45 @@
+
+    @PostMapping
+    public UserDto add(@RequestBody @Validated(UserDto.FullValidated.class) UserDto userDto) {
+//        fullValidateUserDto(userDto);
```
Если код не используется - следует его убрать

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,29 @@
+    String name;
+
+    @Email(groups = {FullValidated.class, PatchValidated.class})
+    @NotNull(groups = FullValidated.class)
```
В данном случае следует использовать так же `@NotBlank` или `@NotEmpty`, так как, если передать строку с почтой в виде "", она будет принята за корректную, так как `@Email` допускает и такой вариант) Поэтому здесь так же хорошо было бы для обновления использовать `@NullableNotBlank(groups = PatchValidated.class)`

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты :) Хорошего дня! Если возникнут вопросы - пиши в слак, буду рад помочь🌆

svlgerasimov committed:
fix: в проверке при поиске предметов заменить isEmpty на isBlank
style: убрать закомментированный код
fix: для поля email в UserDto добавить валидацию NotBlank (с учетом групп валидации)

svlgerasimov committed:
refactor: убрать CustomValidationException, т.к. вся валидация теперь осуществляется через аннотации

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в знакомстве со JPA. Прекрасного вечера🌃
